src/main/java/mate/academy/spring/service/CustomUserDetailsService.java

```diff
@@ -0,0 +1,34 @@
+package mate.academy.spring.service;
+
+import static org.springframework.security.core.userdetails.User.withUsername;
+
+import java.util.Optional;
+import mate.academy.spring.model.User;
+import org.springframework.security.core.userdetails.User.UserBuilder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserService userService;
+
+    public CustomUserDetailsService(UserService userService) {
+        this.userService = userService;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        Optional<User> userOptional = userService.findByEmail(email);
+        if (userOptional.isPresent()) {
+            User user = userOptional.get();
+            UserBuilder userBuilder = withUsername(email);
+            userBuilder.password(user.getPassword());
+            userBuilder.roles(user.getRoles().stream()
+                    .map(r -> r.getName().name()).toArray(String[]::new));
+            return userBuilder.build();
+        }
+        throw new UsernameNotFoundException("User with email " + email + " was not found");
```

Comment:
better to operate with object of User not with Optional
```suggestion
        User user = userService.findByEmail(email).orElseThrow(() -> your exception with message);
         UserBuilder userBuilder = withUsername(email);
         userBuilder.password(user.getPassword());
         userBuilder.roles(user.getRoles().stream()
                    .map(r -> r.getName().name()).toArray(String[]::new));
         return userBuilder.build();
```